cmd/kube-apiserver/app/server.go

```diff
@@ -498,9 +490,7 @@ func buildGenericConfig(
 	}
 	versionedInformers = clientgoinformers.NewSharedInformerFactory(clientgoExternalClient, 10*time.Minute)
 
-	genericConfig.Authentication.Authenticator, genericConfig.OpenAPIConfig.SecurityDefinitions, err = BuildAuthenticator(s, genericConfig.EgressSelector, clientgoExternalClient, versionedInformers)
-	if err != nil {
-		lastErr = fmt.Errorf("invalid authentication config: %v", err)
+	if lastErr = s.Authentication.ApplyTo(genericConfig, clientgoExternalClient, versionedInformers); lastErr != nil {
```

Comment:
This is quite ugly and order dependent.  I'm going to jump down to the definition, but I do know that here I'd prefer to see the `server.RecommendedConfig` used instead of the individual items. You should be able to change the genericConfig to be the recommendedConfig with minimal pain.

Reply:
`server.RecommendedConfig` takes config by value (embeds it), making a shallow copy of the genericConfig to put into a function that is meant to write into it sound kind of error prone, doesn't it? (As a nit you have to create a new client again as RecommendedConfig holds only `restclient.Config`.)

Reply:
> This is quite ugly and order dependent.
It is. Do you think we can do better with the current "ApplyTo" pattern?